change: [M3-8533, M3-8761] - Fix firewall rules table and Replace `react-beautiful-dnd` with `dnd-kit` lib

pmakode-akamai opened the pull request:
## Description 📝
While working on ticket M3-8533 & PR #11109, we decided to implement/use semantic markup for the Firewall rules table to improve consistency and accessibility, rather than applying a styling band-aid to div grids to make it look like our normal tables. The `react-beautiful-dnd` library has made us rely on div(Box) grids instead of our normal tables, which complicates using semantic tables/grid layouts because of the draggable elements. Plus, this library has been deprecated (announcement: https://github.com/atlassian/react-beautiful-dnd/issues/2672).

To solve these problems and make better use of our normal tables, switching to `dnd-kit`(https://dndkit.com/) would be the great idea, which is more popular, supports grid layouts better, and can replace `react-beautiful-dnd`. This will help us use our semantic tables properly and fix the issues with the Firewall rules table in both dark and light modes.

## Changes  🔄
List any change relevant to the reviewer.
- Used Normal tables instead of Div Box grids
- Updated `PolicyRow` 
  - Changed it to have borders on all sides.
  - Changed the alignment of the Select and helper text to be aligned with the Action column for screens < 'md', and with the last column for screens >= 'md'.
- Replaced `react-beautiful-dnd` with `dnd-kit` lib
- Fix `<Table />` styling for `noOverflow` property

## Target release date 🗓️
N/A

## Preview 📷
| Before  | After   |
| ------- | ------- |
| ![Screenshot 2024-10-21 at 6 21 23 PM](https://github.com/user-attachments/assets/dec8e128-2ed0-4296-b1a9-994a1f56277c) | ![Screenshot 2024-10-21 at 6 20 12 PM](https://github.com/user-attachments/assets/7dfb572c-7d74-4137-bded-9ab97cb3c837) |
| ![Screenshot 2024-10-21 at 6 22 28 PM](https://github.com/user-attachments/assets/835833bc-77ae-4121-b902-2a191ab8d5d0) | ![Screenshot 2024-10-21 at 6 23 24 PM](https://github.com/user-attachments/assets/97d5a3e2-4779-4b95-acab-c12af8f27898) |

## How to test 🧪

### Prerequisites
- Clear cache or test in Incognito. 

### Reproduction steps
- Go to any Firewall details page (https://cloud.linode.com/firewalls/948204/rules)
- Observe the broken looking table (in both dark and light mode)

### Verification steps
- Ensure the table is not broken in both dark and light modes.
- Verify table across different screen sizes and ensure it matches our normal table colors.
- Ensure the drag and drop functionality work as expected in different screen sizes.
- Verify all firewall rule actions work as expected.

## As an Author I have considered 🤔

*Check all that apply*

- [ ] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

abailly-akamai commented on packages/manager/package.json:
```diff
@@ -14,6 +14,9 @@
   "dependencies": {
+    "@dnd-kit/core": "^6.1.0",
+    "@dnd-kit/sortable": "^8.0.0",
+    "@dnd-kit/utilities": "^3.2.2",
```
can we also remove `react-beautiful-dnd` since not used anywhere else

pmakode-akamai replied:
Done ✅

abailly-akamai reviewed (commented):
Fantastic work @pmakode-akamai, and really appreciate considering going with a more holistic refactor, including replacing a non-supported library.

Implementation, markup and styling looks great (in both modes) for the most part, leaving some comments that shouldn't take too much extra work to implement.

![Screenshot 2024-10-21 at 10 57 48](https://github.com/user-attachments/assets/d8a23da2-ea58-4132-ac4e-a10649a82453)

![Screenshot 2024-10-21 at 11 03 49](https://github.com/user-attachments/assets/8c431196-e4fa-44c6-badb-431a739c67a3)

abailly-akamai commented:
@pmakode-akamai I believe the e2e failure is relevant here - you may want to take a look at `firewalls/update-firewall.spec.ts`

bnussman-akamai reviewed (commented):
Thanks for working on this!

For whatever reason, the dragging is not nearly as smooth as the previous implementation for me. I did try the dndkit [examples](https://master--5fc05e08a4a65d0021ae0bf2.chromatic.com/?path=/story/presets-sortable-multiple-containers--basic-setup) and they were much smoother. Could this be something wrong on our end?


https://github.com/user-attachments/assets/f517e1a6-5b44-4b7d-897e-ad16249a28e2

I'm also noticing that the table row is can't overflow outside of the table like it did previously when dragging. Not a blocker, but I liked how previously the row was "on top" of everything when dragging.

pmakode-akamai commented:
> For whatever reason, the dragging is not nearly as smooth as the previous implementation for me. I did try the dndkit [examples](https://master--5fc05e08a4a65d0021ae0bf2.chromatic.com/?path=/story/presets-sortable-multiple-containers--basic-setup) and they were much smoother. Could this be something wrong on our end?

Hi @bnussman-akamai, We can definitely add a `transition` (something like in [this](https://docs.dndkit.com/presets/sortable/usesortable#transition-1)) to make the drag animation smoother. However, I'm having some difficulty figuring out how the `dnd-kit` is handling `onDragEnd` in our case.

I'm encountering a similar issue to what was mentioned [here](https://github.com/clauderic/dnd-kit/issues/833) when I try to add the transition. Looks like `dnd-kit` works fine with local state but not so much with async state (as per this [comment](https://github.com/clauderic/dnd-kit/issues/833#issuecomment-2251216009)).  In our case, the way we are reordering `onDragEnd` is not straight forward. 

I will check how we can add a transition in our case somehow.

> I'm also noticing that the table row is can't overflow outside of the table like it did previously when dragging. Not a blocker, but I liked how previously the row was "on top" of everything when dragging.

I agree, the `DndContext` is currently applied to the table, which is why the table row can’t overflow outside of it like it did previously. We can explore adjustments to allow the row to appear 'on top' during dragging.

pmakode-akamai commented:
@abailly-akamai @bnussman-akamai 

> For whatever reason, the dragging is not nearly as smooth as the previous implementation for me. I did try the dndkit [examples](https://master--5fc05e08a4a65d0021ae0bf2.chromatic.com/?path=/story/presets-sortable-multiple-containers--basic-setup) and they were much smoother. Could this be something wrong on our end?

update: The smoothness while dragging rows has been improved! ✅

> I'm also noticing that the table row is can't overflow outside of the table like it did previously when dragging. Not a blocker, but I liked how previously the row was "on top" of everything when dragging.

To allow dragging over the top of the table/across the screen, we can simply add a `noOverflow={true}` prop to the `<Table />` component (as overflow is not allowing us to drag over the top of the table/across the screen). However, this changes the styling of the header due to how `noOverflow` is implemented in our `<Table />` component.

- With `noOverflow={true}`
![Screenshot 2024-10-23 at 9 30 57 PM](https://github.com/user-attachments/assets/3fbcabf8-71f9-4cd4-8457-add67e9fd3b4)

I’d like to confirm if we want to apply this header styling with noOverflow={true} for the FirewallRulesTable. Additionally, should we maintain the overflow settings for Firewall Rules Table, considering that the version in production does not have overflow?

abailly-akamai reviewed (approved):
Great work - thanks for the contribution 👍

bnussman-akamai commented:
In production I'm able to re-order rules with the keyboard (using `space` then `up` and `down` arrow keys). In this PR I don't think I can do that, but I may be missing something. Is this something we can accomplish with dndkit?

abailly-akamai commented:
@bnussman-akamai 

The team seems to have put great care into accessibility for this component, which is great - https://docs.dndkit.com/guides/accessibility#keyboard-support 

This implementation was missing the keyboard sensors, which surprise me isn't a default. Anyway, added them. You experience may differ slightly from the previous plugin tho

hkhalil-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Rules/FirewallRuleTable.tsx:
```diff
@@ -278,88 +290,101 @@ const FirewallRuleTableRow = React.memo((props: FirewallRuleTableRowProps) => {
+    },
+    cursor: isDragging ? 'grabbing' : 'grab',
+    position: 'relative',
+    transform: CSS.Transform.toString(transform),
```
There is a weird scaling effect when rows have different heights:

https://github.com/user-attachments/assets/cf8d3d3d-4dc7-4ada-98ec-5e393fbfe545

The following solves it, taken from [here](https://github.com/clauderic/dnd-kit/issues/117):

```suggestion
    transform: CSS.Translate.toString(transform),
```

pmakode-akamai replied:
good catch 👍

hkhalil-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Rules/FirewallRuleTable.tsx:
```diff
@@ -115,16 +128,44 @@ export const FirewallRuleTable = (props: FirewallRuleTableProps) => {
+      },
+    }),
+    useSensor(TouchSensor),
+    useSensor(KeyboardSensor)
```
I don't think the `KeyboardSensor` achieves the effect we're looking for. It moves items in the keyed direction by 25px, which can make it difficult to actually change the place of the item. It allows moving the item left and right and also scrolls the page.

https://github.com/user-attachments/assets/73177091-292a-4e03-aa5a-b141f9e4b33c

I can't find an option within dnd-kit that accomplishes the prior behavior. My recommendation is to setup a custom key handler that calls `triggerReorder` as well as `e.preventDefault()` to prevent the scrolling issue.

hkhalil-akamai replied:
This gets closer but I'm still seeing some weird behavior when the page has vertical scrolling enabled. It's hard to describe but you can see it for yourself by `Ctrl`-`+`'ing in until the page overflows.

pmakode-akamai replied:
@abailly-akamai @hkhalil-akamai I tried different ways and even used `e.preventDefault()` to prevent window scrolling (when the browser has vertical scroll enabled) while the `KeyboardSensor` is active. However, this didn't work due to how the `dnd-kit` KeyboardSensor is implemented, as it adds default scrolling behavior. After extensive digging 🥲, I found that the `handleKeyDown` method from the `KeyboardSensor` does not allow us to override the scrolling behavior when using keyboard keys. As a result, I customized the dnd-kit library [code](https://github.com/clauderic/dnd-kit/blob/master/packages/core/src/sensors/keyboard/KeyboardSensor.ts) itself to meet our requirements 😅, and some how I was able to prevent the scroll behavior when using keyboard keys. ✅

Could you take a look?

https://github.com/linode/manager/pull/11127/commits/a839fef8ebd8d08ae888ca682ab68cda2472a3ba

hkhalil-akamai replied:
Great digging, this works exactly like I would expect it to! 👏🏽🎉

pmakode-akamai commented on packages/manager/src/components/Table/Table.styles.ts:
```diff
@@ -14,27 +14,26 @@ export const StyledTableWrapper = styled('div', {
+      fontFamily: theme.font.bold,
+      padding: '10px 15px',
     },
-    overflowX: 'auto',
-    overflowY: 'hidden',
```
@abailly-akamai 

> * I looked at a couple components that had this property declared and did not see any regression. Not sure what the intent was when adding it, but since it did the opposite it's worth checking

Thank you for the changes, but I wonder if the intent was to enable a horizontal scrollbar for all tables by default, without the `noOverflow` property (for smaller screens or at time when the table content tend to overflow)? 

| On Prod  | On this Branch with these changes  | 
| ------- | ------- | 
| ![Screenshot 2024-10-24 at 4 25 16 PM](https://github.com/user-attachments/assets/2df54b0f-bff3-4abc-b128-60519937c258) | ![Screenshot 2024-10-24 at 4 26 30 PM](https://github.com/user-attachments/assets/495cb573-c8e4-49eb-802e-3064d6519698) |

If the plan is to apply `overflowX: auto` to all our normal tables by default (incase), we need to ensure this doesn't affect the Firewall Rules Table. As, we want to allow row dragging over the top in this case, which won’t be possible if overflowX: auto is enabled. With some changes here; we may need to add `noOverflow` property specific to the Firewalls Rules table.

abailly-akamai replied:
Yeah sorry I am working too fast on this one - I'll revert that part but keep the styling fix

hkhalil-akamai reviewed (approved):
Great work @pmakode-akamai @abailly-akamai! 🎉

abailly-akamai reviewed (approved):
Thanks @pmakode-akamai - great work!

While I am not particularly enthusiastic about overriding the library this way (it makes things more subject to breaking during future updates), it also solves the problem. I was thinking we could have merged as it was before this addition and add an issue on their repo to address it in a future release, since this keyboard interaction is very much aimed to facilitate accessibility, not necessarily "look good". That being said, I also like the fact we're keeping parity with the previous implementation.

All that being said, thanks for addressing all feedback and finding clever workarounds!

nit: a unit test for `CustomKeyboardSensor.ts` could be a decent addition to add safety to this customization

bnussman-akamai reviewed (commented):
Keyboard navigation seems to be working similarly to the old implementation now. The focus state doesn't seem as obvious on this branch as it does in develop. 

Should we have the row change color when it is being focused/moved via the keyboard?

pmakode-akamai commented:
> Keyboard navigation seems to be working similarly to the old implementation now. The focus state doesn't seem as obvious on this branch as it does in develop.
> 
> Should we have the row change color when it is being focused/moved via the keyboard?

It looks like there is a focus issue in Prod where the focus state resets when moving the row via keyboard - it resets to the initial position which is distracting. 

**Prod Issue:**

https://github.com/user-attachments/assets/c6a22969-643a-43bd-9c2b-f4479e2d3209


Commit underneath fixes the above issue in this branch by ensuring the draggable element remains focused until it is dropped. Additionally, I’ve modified the focus styling (outline dashed), which is bit different but similar to our prod version. If we have any plan to change the row color for focus (in case), we can definitely implement that once we finalize the color we want to use. For now, I’ve only improved the dashed border and fixed the focused state issue I saw in Prod.

fbfdcdc6af4afed377589759ac7c801c1155e1a8

Could you check and confirm?

bnussman-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Rules/FirewallRuleTable.tsx:
```diff
@@ -112,16 +131,46 @@ export const FirewallRuleTable = (props: FirewallRuleTableProps) => {
+      },
+    }),
+    useSensor(TouchSensor),
+    useSensor(CustomKeyboardSensor, {
+      coordinateGetter: sortableKeyboardCoordinates,
+    })
```
I'd almost just be in-favor of using the default KeyboardSensor just so we don't have the maintenance burden of a custom one. Do the benefits of the custom one outweigh convenience of the built in one? 

I tried the default one and it seemed to work "good enough"

abailly-akamai replied:
I agree ++

hkhalil-akamai replied:
I would argue the opposite. The default seems to work fine in default viewports but breaks down in small viewports when a scrollbar is introduced.

See the attached video; when at default zoom, keyboard navigation works as expected but when zooming in, we see that some keystrokes are not recognized or have no effect:


https://github.com/user-attachments/assets/f76f18e7-6b56-4fab-a3ad-7d6cc94753af

abailly-akamai replied:
Isn't this a super edge case? (keyboard navigation zoomed in/on mobile)

bnussman-akamai replied:
Hmm. I'm not able to reproduce that using

```typescript
    useSensor(KeyboardSensor, {
      coordinateGetter: sortableKeyboardCoordinates,
    })
```

hkhalil-akamai replied:
@abailly-akamai good point. @bnussman-akamai maybe a Safari vs Chrome thing?

bnussman-akamai replied:
Yeah, could be. First thing next week I'll dive a bit deeper so we can reach a consensus here

pmakode-akamai replied:
I have just customized the [`handleKeyDown`](https://github.com/clauderic/dnd-kit/blob/e2a1776d0de657669192d3cfd1558e91905b5fad/packages/core/src/sensors/keyboard/KeyboardSensor.ts#L80) method of the default KeyboardSensor, specifically removing the scrolling logic from [L126-L236](https://github.com/clauderic/dnd-kit/blob/e2a1776d0de657669192d3cfd1558e91905b5fad/packages/core/src/sensors/keyboard/KeyboardSensor.ts#L126-L236). This prevents scrolling issues on smaller devices, which in turn resolves the problem of keystrokes not being recognized or having no effect on smaller screens.

Additionally, we could extend/add some logic (like focus styling and other keyboard accessibility improvements) within the CustomKeyboardSensor in other methods, rather than adding these on the component level, which seems less straightforward.

Given this, should we stick with the default sensor, or continue using the custom sensor, considering this super edge case with smaller devices?

bnussman-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Rules/FirewallRuleTable.tsx:
```diff
@@ -275,88 +305,101 @@ const FirewallRuleTableRow = React.memo((props: FirewallRuleTableRowProps) => {
+      // in order to prevent scrolling on mobile devices.
+      // refer to https://docs.dndkit.com/api-documentation/sensors/pointer#touch-action
+      touchAction: 'none',
+    },
```
I'm not sure if this is recommended from an accessibility perspective, but I'm finding that it is helpful to make it more obvious that a row is focused

```suggestion
    },
    ':focus': {
      backgroundColor: 'lightblue',
    },
```

pmakode-akamai replied:
`lightblue` color looks a bit odd in dark mode and makes the row text unreadable. Is there a more theme-sensitive background color we could use for focus, other than the one we used for edit, clone, and delete statuses (to avoid color conflict)?
![Screenshot 2024-11-11 at 3 32 15 PM](https://github.com/user-attachments/assets/de54e898-3f25-43eb-90e5-ea8ab6a3bc8e)

pmakode-akamai commented:
@bnussman-akamai Are we good to merge this PR? thanks!!

abailly-akamai commented:
@pmakode-akamai have you confirmed the failing tests are unrelated?

pmakode-akamai commented:
> @pmakode-akamai have you confirmed the failing tests are unrelated?

@abailly-akamai yea, this should be unrelated

confirming all checks for these 2 specs pass locally. I think all checks should pass on re-run.

<details>

![Screenshot 2024-11-15 at 11 23 00 PM](https://github.com/user-attachments/assets/bd75c1ba-7cfc-4a9b-9107-7987d39fdbbe)

![Screenshot 2024-11-15 at 11 24 59 PM](https://github.com/user-attachments/assets/646bb1b2-762e-44bf-bf0e-cf723ed61d4c)
</details>
